### PR TITLE
Overwriting of mutated props in constructor

### PR DESCRIPTION
**Pull request by drdelambre**
When rendering the class ```Bar``` from the following example, the ```Bar``` instance has the moo prop set as 'BOO' and not the expected 'BOO!' because the props passed to the constructor are overwritten with their initial state in the chance they are undefined. This PR just ignores that logic if props is defined.

```Javascript
class Foo extends React.Component {
  constructor(props) {
    var _props = { moo: new.target.defaultProps.moo + '!' }
    super(_props);
  }
}
Foo.defaultProps = { moo: 'MOO' };

class Bar extends Foo {}
Bar.defaultProps = { moo: 'BOO' };
```

### CODE REVIEW

**Comment by jimfb** [comment]
Yeah, this behavior does seem more correct to me, and the PR looks good.

I'll leave the PR open for a day or two so everyone gets a chance to take a look and raise any red flags before we merge.

File: src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js [hunk 1]
```diff
@@ -1241,4 +1241,53 @@ describe('ReactCompositeComponent', function() {
...
+    function Bar() {
+      Foo.call(this);
+    }
+
+    Bar.prototype = Object.create(Foo.prototype, {
+      constructor: {
+        value: Bar,
+      },
+    });
+    Bar.defaultProps = { idx: 'bar' };
```
Comment (zpao) [inline src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js]: Let's just use `class` syntax. We transform our tests.

**Comment by zpao** [comment]
cc @sebmarkbage What level of support are we giving this inheritance pattern?

File: src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js (hunk 1 above)
Comment (jimfb) [inline src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js]: Yeah, I agree, that would be better.  @drdelambre Would you mind updating this diff?

**Comment by jimfb** [comment]
I got the idea from other threads that we are "strongly encouraging composition over inheritance", but will generally try to make composition work.  This diff makes sense in the interest of correctness, even for components that directly extend `React.Component`.

**Comment by jimfb** [comment]
Ok, looks good!  Thanks @drdelambre

**Comment by sebmarkbage** [comment]
No, we should revert this. This pattern only works for the initial set of props. It doesn't work for updates. So you end up violating an important invariant for React.

**Comment by drdelambre** [comment]
I was saving that for a more heated discussion in another PR where the reason for not setting ```nextProps = inst.props``` when ```prevParentElement === nextParentElement``` in ```updateComponent``` is reveled, as there's no tests for the condition it's comments warn about and I can't trace the edge case.

**Comment by jimfb** [comment]
I was in the process of filing a bug to add a warning as you were writing this comment.

**Comment by drdelambre** [comment]
Is there a thread on why ```getDefaultProps```isn't included in ```ReactComponent``` for es6 classes? It would solve these edge cases I'm experiencing (default instances in an object hierarchy when a property isn't defined) without restructuring internal code.

**Comment by jimfb** [comment]
When a property isn't defined?  `getDefaultProps` happens before props is available: https://facebook.github.io/react/docs/component-specs.html#getdefaultprops

**Comment by drdelambre** [comment]
If a default property is a basic type, totally works fine, but if a default property needs to be a class instance, then the class needs to be initialized upon object construction or else all instances of the component are sharing the same instance. Is the correct nomenclature defaultProperty property?

**Comment by jimfb** [comment]
Right, `getDefaultProps` is cached anyway.  Even with the function variant, if you return a complex type, all instances of the component will share the same instance.

You can think of getDefaultProps as a static function on a Java class, if that's helpful to you.  Or as an immutable public static final global.  It is used to specify the "defaults" and is completely independent of instances.  It is generally used to specify intelligent defaults for a component configuration, not to initialize initial internal state (that's what `getInitialState()` does).

Props and state are intentionally separate.  See: https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html

**Comment by drdelambre** [comment]
oh ok. I thought ```getDefaultProps``` was a magic bullet that was denied es6, but I see now why it was removed. I'll get to refactoring how state is addressed then! Thanks for your help @jimfb

**Comment by sebmarkbage** [comment]
The `getDefaultProps` name is just there for legacy reasons. Could be `statics: { defaultProps: ... } }`

**Pull request merged**